use acp client tools if available

haoxiangliew opened the pull request:
Delegates file `reads`, `edits`, and `writes` to the ACP Client, if the ACP Client (such as Zed) supports these modifications.

Brings `opencode acp` more in-line with `@zed-industries/claude-code-acp`
- Contributes to #216 :)

<img width="2652" height="1500" alt="acp" src="https://github.com/user-attachments/assets/d2ecd3c8-d68b-4612-9ab4-bdae40fba974" />

haoxiangliew committed:
init

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

rekram1-node commented:
@haoxiangliew i was thinking about doing this, we should probably disable the opencode write tools then if the ACP provided ones are avialable no?

haoxiangliew commented on packages/opencode/src/acp/agent.ts:
```diff
@@ -601,6 +634,7 @@ export namespace ACP {
           },
           parts,
           agent,
+          tools,
```
> @haoxiangliew i was thinking about doing this, we should probably disable the opencode write tools then if the ACP provided ones are avialable no?

@rekram1-node tools should be disabled here if `hasACPTools`, if there's a better place to disable them lmk

rekram1-node replied:
okay cool

rekram1-node commented:
to work with our tools & prompt I think we should just make the acp tools be called the exact same thing with the same prompts ideally. 

I think you could import the existing tools, and override / copy from them so that you only change the execute portions

also you need to remember to keep the permissions system in place too

haoxiangliew committed:
refactor ACP tools

take PR feedback into account:
- Import existing tools and only modify `execute()`
- No prompt change needed, reads, edits, and writes are executed as-is
with a different interface
- Keep permissions

refactor ACP tools

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew commented:
> to work with our tools & prompt I think we should just make the acp tools be called the exact same thing with the same prompts ideally.
> 
> I think you could import the existing tools, and override / copy from them so that you only change the execute portions
> 
> also you need to remember to keep the permissions system in place too

Thanks for the feedback @rekram1-node! Implemented in 25e6929b02070634a5b9a15048389728c327919b

haoxiangliew committed:
revert changes to readme

haoxiangliew committed:
refactor ACP tools

haoxiangliew committed:
we don't need ACP for "does a file exist"

we don't need ACP for "does a file exist"

haoxiangliew committed:
remove unused import

remove unused import

remove unused import

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
assure both capabilities are there

haoxiangliew commented:
> to work with our tools & prompt I think we should just make the acp tools be called the exact same thing with the same prompts ideally.
> 
> I think you could import the existing tools, and override / copy from them so that you only change the execute portions
> 
> also you need to remember to keep the permissions system in place too

@rekram1-node 

went thru a couple iterations on the design to get this done as cleanly as possible:
- wanted to preserve the read,write,edit tools as-is instead of duplicating them across to acp
- \+ no need to disable the native tools when using acp

lmk what u think :+1:

rekram1-node commented on packages/opencode/src/acp/session.ts:
```diff
@@ -2,14 +2,21 @@ import type { McpServer } from "@agentclientprotocol/sdk"
+import type { ACP } from "./agent"
 
 export class ACPSessionManager {
+  private static registry = new Map<string, ACPSessionManager>()
```
I feel like the acp tools could just have an init function similar to how other tools do so we wouldn't have to do this and instead each tool could have direct access to the session manager without this "global"

rekram1-node commented on packages/opencode/src/acp/agent.ts:
```diff
@@ -361,9 +368,21 @@ export namespace ACP {
+        log.info("creating_session", {
+          mcpServers: params.mcpServers.length,
+          hasACPTools,
+        })
```
Given there are 2 different ACP tools, just for compatability purposes, do you think we can do:

clientSupportsRead
clientSupportsWrite

instead of tying both up under "hasACPTools"

and then only override w/ acp tool if the client supports that tool

rekram1-node commented on packages/opencode/src/tool/edit.ts:
```diff
@@ -73,7 +76,15 @@ export const EditTool = Tool.define("edit", {
+            content: params.newString,
+          })
+        } else {
+          await Bun.write(filePath, params.newString)
```
Ideally we dont have to reimplement entire edit tool if the client doesnt support the edit functionality

I think we could just do a check, if client supports write use ACP write and remove the native write/edit tools
else: use the native write/edit tools

rekram1-node commented:
left a couple comments here. Also I'm gonna merge dev into your branch there was a bug in one of the tests that causes CI to run for a longgg time

rekram1-node committed:
Merge branch 'dev' into zed-acp-fixes

rekram1-node committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
address PR

haoxiangliew commented on packages/opencode/src/tool/registry.ts:
```diff
@@ -86,8 +87,8 @@ export namespace ToolRegistry {
       TodoWriteTool,
       TodoReadTool,
       TaskTool,
-      ...custom,
-    ]
+    ].filter((tool) => !customIds.has(tool.id))
+    return [...nativeTools, ...custom]
   }
```
if a custom tool uses the same id as a native tool, use the custom tool

haoxiangliew committed:
address PR

haoxiangliew committed:
cleanup

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
cleanup

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
refactor merge after 'dev'

haoxiangliew committed:
update

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
cleanup

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

actions-user committed:
chore: format code

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
fix after merge

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

rekram1-node commented:
I've mentioned it in discord and when I get the chance I can update this pr but we don't want to bleed the domains, the bash, edit, etc tools should not import anything ACP related at all. 

ACP should be able to do all this entirely in its own domain just importing from the others and overriding things if necessary.

I think this requires a bit more work

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
rework for domain

actions-user committed:
chore: format code

haoxiangliew committed:
update

actions-user committed:
chore: format code

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

actions-user committed:
chore: format code

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

actions-user committed:
Update Nix flake.lock and hashes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

actions-user committed:
Update Nix flake.lock and hashes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
bump

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
Merge branch 'dev' into zed-acp-fixes

haoxiangliew committed:
update
